File: blackjack/csharp/Hand.cs
using System;
using System.Collections.Generic;

public class Hand
{
    private List<Card> cards;
    private int hard;
    private int soft;
    private bool revealed;

    // construct the hand using two cards
    public Hand(Card down, Card up) {
        // create the cards list
        cards = new List<Card>();

        // initialize the scores
        hard = 0;
        soft = 0;
        revealed = false;

        // add the cards
        hit(down);
        hit(up);
    }


    // Accessor for revealed status
    public bool getRevealed() {
        return revealed;
    }


    // Mutator for revealed status
    public void setRevealed(bool revealed) {
        this.revealed = revealed;
    }


    // Acessor for the best point value
    public int getPoint() {
        if(soft != 0) {
            return soft;
        }

        return hard;
    }


    // get the hard value (with aces 1)
    public int getHard() {
        return hard;
    }


    // get soft value (zero if soft busts or no ace) 
    public int getSoft() {
        return soft;
    }


    // receive a hit
    public void hit(Card card) {
        int value; // numeric value of the current card

        // add the card to the 
        cards.Add(card);

        // get the card's value
        switch(card.getValue()) {
            case "A":
                value = 1;
                break;
            case "J":
            case "Q":
            case "K":
                value = 10;
                break;
            default:
                value = int.Parse(card.getValue());
                break;
        }

        // add to the hard score
        hard += value;

        // handle the soft score
        if(soft != 0 || value == 1) {
            soft = hard + 10;
        }

        // reset a busted soft
        if(soft > 21) {
            soft = 0;
        }
    }


    // enumerate the cards in the hand
    public IEnumerable<Card> getCards() {
        foreach(Card card in cards) {
            yield return card;
        }
    }


    // convert the hand to a string for printing
    public override string ToString() {
        string result = "";
        bool first = true;

        foreach(Card card in getCards()) {
            if(!first || getRevealed()) {
                result += String.Format("{0} ", card);
            } else {
                result += "### ";
            }

            // toggle first off and move on
            first = false; 
        }

        if(getRevealed()) {
            result += "(";
            result += getHard();
            int soft = getSoft();
            if(soft != 0) {
                result += " / " + soft;
            }
            result += ")";
        }

        return result;
    }
}

File: blackjack/ruby/hand.rb
class Hand
  #initialize the hand with two cards
  def initialize(down, up)
    # point values start at zero
    @hard = 0
    @soft = 0

    # create our hand, and hide it by default
    @cards = []
    @revealed = false

    # add the cards
    hit(down)
    hit(up)
  end


  #get the best point value
  def point
    if @soft != 0
      return @soft
    else
      return @hard
    end
  end


  # receive a hit
  def hit(card)
    # add the card to the hand
    @cards.append(card)

    # get the numeric value of the card
    case card.value
    when "A"
      value = 1
    when "J"
      value = 10
    when "Q"
      value = 10
    when "K"
      value = 10
    else
      value = card.value.to_i
    end

    # add to the hard score
    @hard += value

    #handle the soft score
    if @soft != 0 or value == 1
      @soft = @hard + 10
    end

    # reset a busted soft
    if @soft > 21
      @soft = 0
    end

    return nil
  end


  # convert to a string suitable for printing
  def to_s
    # get our enumerator
    enum = to_enum

    # initialize the output
    show=revealed
    result =""

    # handle the rest of the cards
    enum.each do |card|
      if show
        result += "%s " % [card.to_s]
      else
        result += "### "
      end

      # show all subseqent cards
      show = true
    end

    # add in the point value
    if revealed
      if soft != 0 
        result += "(#{@hard} / #{@soft})"
      else
        result += "(#{@hard})"
      end
    end

    return result
  end

  
  # provide a method to enumerate the hand
  def to_enum
    @cards.to_enum
  end


  # accessors
  attr_accessor :revealed
  attr_reader :hard
  attr_reader :soft
end
